plutus-core/src/Language/PlutusCore/Flat.hs

```diff
@@ -0,0 +1,356 @@
+encodeKind = eBits kindTagWidth
+
+decodeKind :: Get Word8
+decodeKind = dBEBits8 kindTagWidth
+
+instance Flat ann => Flat (Kind ann) where
+    encode = \case
+        Type ann           -> encodeKind 0 <> encode ann
+        KindArrow ann k k' -> encodeKind 1 <> encode ann <> encode k  <> encode k'
+
+    decode = go =<< decodeKind
+        where go 0 = Type      <$> decode
+              go 1 = KindArrow <$> decode <*> decode <*> decode
+              go _ = fail "Failed to decode Kind ()"
+
+-- | Use 3 bits to encode type tags.
+typeTagWith :: NumBits
+typeTagWidth = 3
+
+encodeType :: Word8 -> Encoding
+encodeType = eBits typeTagWidth
+
+decodeType :: Get Word8
+decodeType = dBEBits8 typeTagWidth
+
+instance (Closed uni, Flat ann, Flat tyname) => Flat (Type tyname uni ann) where
+    encode = \case
+        TyVar     ann tn      -> encodeType 0 <> encode ann <> encode tn
+        TyFun     ann t t'    -> encodeType 1 <> encode ann <> encode t   <> encode t'
+        TyIFix    ann pat arg -> encodeType 2 <> encode ann <> encode pat <> encode arg
+        TyForall  ann tn k t  -> encodeType 3 <> encode ann <> encode tn  <> encode k <> encode t
+        TyBuiltin ann con     -> encodeType 4 <> encode ann <> encode con
+        TyLam     ann n k t   -> encodeType 5 <> encode ann <> encode n   <> encode k <> encode t
+        TyApp     ann t t'    -> encodeType 6 <> encode ann <> encode t   <> encode t'
+
+    decode = go =<< decodeType
+        where go 0 = TyVar     <$> decode <*> decode
+              go 1 = TyFun     <$> decode <*> decode <*> decode
+              go 2 = TyIFix    <$> decode <*> decode <*> decode
+              go 3 = TyForall  <$> decode <*> decode <*> decode <*> decode
+              go 4 = TyBuiltin <$> decode <*> decode
+              go 5 = TyLam     <$> decode <*> decode <*> decode <*> decode
+              go 6 = TyApp     <$> decode <*> decode <*> decode
+              go _ = fail "Failed to decode Type TyName ()"
+
+    size tm sz = 3 + sz + case tm of
+        TyVar     ann tn      -> getSize ann + getSize tn
+        TyFun     ann t t'    -> getSize ann + getSize t   + getSize t'
+        TyIFix    ann pat arg -> getSize ann + getSize pat + getSize arg
+        TyForall  ann tn k t  -> getSize ann + getSize tn  + getSize k + getSize t
+        TyBuiltin ann con     -> getSize ann + getSize con
+        TyLam     ann n k t   -> getSize ann + getSize n   + getSize k + getSize t
+        TyApp     ann t t'    -> getSize ann + getSize t   + getSize t'
+
+instance Flat DynamicBuiltinName where
+    encode (DynamicBuiltinName name) = encode name
+    decode = DynamicBuiltinName <$> decode
+
+-- | Using 1 bit to encode builtin name tag
+builtinNameTagWidth :: NumBits
+builtinNameTagWidth = 1
+
+encodeBuiltinName :: Word8 -> Encoding
+encodeBuiltinName = eBits builtinNameTagWidth
+
+decodeBuiltinName :: Get Word8
+decodeBuiltinName = dBEBits8 builtinNameTagWidth
+
+instance Flat BuiltinName where
```

Comment:
size?

Reply:
Added size.

Reply:
I guess it seemed sensible for the cases where we aren't using the generic instances and we're hand-coding in the tag alternatives. But maybe the generic instance is still fine there: if we make a mistake in how we allocate the tags it'll break other things.